src/os/bluestore/BlueFSCache.cc

```diff
@@ -0,0 +1,203 @@
+#include <functional>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <shared_mutex>
+#include <unordered_map>
+#include "common/errno.h"
+#include "common/dout.h"
+#include "common/debug.h"
+#include "include/buffer_fwd.h"
+#include "common/ceph_context.h"
+#include "include/denc.h"
+#include "tools/cephfs/EventOutput.h"
+
+
+#define dout_context cct
+#define dout_subsys ceph_subsys_bluefs
+#undef dout_prefix
+#define dout_prefix *_dout << "bluefs "
+
+LRUCache::LRUCache(size_t cache_capacity, size_t evict_bytes, CephContext* cct)
+        : cache_capacity(cache_capacity), evict_bytes(evict_bytes), current_bytes(0), cct(cct) {}
+
+cache_bufferlist LRUCache::get_nearest_offset(uint64_t inode, uint64_t offset,
+                                              uint64_t *buffer_offset) {
+
+  dout(20) << __func__ << " inode " << inode << std::hex << " offset 0x" << offset << std::dec << dendl;
+  std::shared_lock lock(mutex);
+  inode_offset_pair key = {inode, offset};
+  auto it = kv.find(key);
+  if (it != kv.end()) {
+    dout(20) << __func__ << " inode " << inode << std::hex << " offset 0x" << offset << " exists in cache" << std::dec << dendl;
+    lock.unlock();
+    *buffer_offset = 0;
+    update_recency(key);
+    return it->second;
+  }
+  it = kv.upper_bound({inode, offset});
+  --it;
+  if (it != kv.end() && it->first.first == inode &&
+      (offset >= it->first.second &&
+       offset < (it->first.second + it->second->length()))) {
+    key = it->first;
+    lock.unlock();
+    update_recency(key);
+    *buffer_offset = offset - it->first.second;
+    dout(20) << __func__ << " inode " << inode << std::hex << " offset 0x" << offset << " exists within offset 0x" << it->first.second  << " at buffer offset 0x" << buffer_offset << std::dec <<  dendl;
+    return it->second;
+  }
+  return NULL;
+}
+
+int LRUCache::put(uint64_t inode, uint64_t offset, cache_bufferlist value) {
+  size_t new_size = value->length();
+
+  std::unique_lock lock(mutex);
+
+  auto it = kv.find({inode, offset});
+
+  if (it != kv.end()) {
+    size_t old_size = it->second->length();
+    // if the bufferlist value is less than the existing bufferlist, no need to
+    // update the map
+    if (new_size < old_size) {
+      dout(20) << __func__ << " inode " << inode << std::hex << " 0x" << offset << "~" << new_size << " is less than the already existing bufferlist size 0x" << old_size << std::dec << dendl;
+      return 0;
+    }
+
+    if (new_size > cache_capacity) {
+      dout(20) << __func__ << " inode " << inode << std::hex << " 0x" << offset << "~" << new_size << " is greater than the cache capacity 0x" << cache_capacity << std::dec << dendl;
+      return -EINVAL;
+    }
+
+    if (current_bytes + new_size - old_size > cache_capacity) {
+      dout(20) << __func__ << " inode " << inode << std::hex << " 0x" << offset << "~" << new_size << " evict cache necessary" << std::dec << dendl;
+      evict(std::max(evict_bytes, new_size));
```

Comment:
I'd suggest moving the std::max() logic into evict itself rather than having it in put.